File: android/feature/rewards/src/main/java/com/enterprise/feature/rewards/domain/usecases/RewardsUseCases.kt
package com.enterprise.feature.rewards.domain.usecases

import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.entities.Reward
import com.enterprise.core.domain.entities.RewardsAccount
import com.enterprise.core.domain.repositories.RewardsRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

/** Gets rewards list as a reactive stream. */
class GetRewardsUseCase @Inject constructor(
    private val rewardsRepository: RewardsRepository
) {
    operator fun invoke(): Flow<DomainResult<List<Reward>>> {
        return rewardsRepository.getRewards()
    }
}

/** Redeems a reward by ID. */
class RedeemRewardUseCase @Inject constructor(
    private val rewardsRepository: RewardsRepository
) {
    suspend operator fun invoke(rewardId: String): DomainResult<RewardsAccount> {
        return rewardsRepository.redeemReward(rewardId)
    }
}

File: ios/EnterpriseMobileApp/Sources/Feature/Rewards/Domain/UseCases/RewardsUseCases.swift
import Foundation
import CoreDomain

/// Use case for fetching available rewards.
public struct GetRewardsUseCase: Sendable {
    private let rewardsRepository: RewardsRepository

    public init(rewardsRepository: RewardsRepository) {
        self.rewardsRepository = rewardsRepository
    }

    public func execute() -> AsyncStream<DomainResult<[Reward]>> {
        rewardsRepository.getRewards()
    }
}

/// Use case for redeeming a reward.
public struct RedeemRewardUseCase: Sendable {
    private let rewardsRepository: RewardsRepository

    public init(rewardsRepository: RewardsRepository) {
        self.rewardsRepository = rewardsRepository
    }

    public func execute(rewardId: String) async -> DomainResult<RewardsAccount> {
        await rewardsRepository.redeemReward(rewardId: rewardId)
    }
}
